jax/_src/scipy/eigh.py

```diff
@@ -0,0 +1,223 @@
+  P = -0.5 * _fill_diagonal(U, U.diagonal() - 1.)
+  rank = jnp.round(jnp.trace(P)).astype(jnp.int32)
+  rank = int(rank)
+  return _split_spectrum_jittable(P, H, V0, rank, precision)
+
+
+def _eigh_work(
+  H, V=None, precision=lax.Precision.HIGHEST, termination_size=128):
+  """ The main work loop performing the symmetric eigendecomposition of H.
+  Each step recursively computes a projector into the space of eigenvalues
+  above jnp.mean(jnp.diag(H)). The result of the projections into and out of
+  that space, along with the isometries accomplishing these, are then computed.
+  This is performed recursively until the projections have size 1, and thus
+  store an eigenvalue of the original input; the corresponding isometry is
+  the related eigenvector. The results are then composed.
+
+  This function cannot be Jitted because the internal split_spectrum cannot
+  be.
+
+  Args:
+    H: The Hermitian input.
+    V: Stores the isometries projecting H into its subspaces.
+    precision: The matmul precision.
+
+  Returns:
+    H, V: The result of the projection.
+  """
+  if H.shape[0] <= termination_size:
+    evals, evecs = jnp.linalg.eigh(H)
+    if V is not None:
+      evecs = jnp.dot(V, evecs, precision=precision)
+    return evals, evecs 
+
+  split_point = jnp.median(jnp.diag(H)) # TODO: Improve this?
+  Hm, Vm, Hp, Vp = split_spectrum(H, split_point, V0=V, precision=precision)
+  Hm, Vm = _eigh_work(
+    Hm, V=Vm, precision=precision, termination_size=termination_size)
+  Hp, Vp = _eigh_work(
+    Hp, V=Vp, precision=precision, termination_size=termination_size)
+
+  if Hm.ndim != 1 or Hp.ndim != 1:
+    raise ValueError(f"One of Hm.ndim={Hm.ndim} or Hp.ndim={Hp.ndim} != 1 ",
+                      "indicating recursion terminated unexpectedly.")
+
+  evals = jnp.hstack((Hm, Hp))
+  evecs = jnp.hstack((Vm, Vp))
+  return evals, evecs
+
+
+def eigh(H, precision=lax.Precision.HIGHEST, symmetrize=True, 
+  termination_size=128):
+  """ Computes the eigendecomposition of the symmetric/Hermitian matrix H.
+
+  Args:
+    H: The `n x n` Hermitian input.
+    precision: The matmul precision.
+    symmetrize: If True, `0.5 * (H + H.conj().T)` rather than `H` is used.
+    termination_size: Recursion ends once the blocks reach this linear size.
+  Returns:
+    vals: The `n` eigenvalues of `H`, sorted from lowest to higest.
+    vecs: A unitary matrix such that `vecs[:, i]` is a normalized eigenvector
+      of `H` corresponding to `vals[i]`. We have `H @ vecs = vals * vecs` up
+      to numerical error.
+  """
+  nrows, ncols = H.shape
+  if nrows != ncols:
+    raise TypeError(f"Input H of shape {H.shape} must be square.")
+
+  if ncols <= termination_size:
+    return jnp.linalg.eigh(H)
+
+  evals, evecs = _eigh_work(H, precision=precision)
+  sort_idxs = jnp.argsort(evals)
+  evals = evals[sort_idxs]
+  evecs = evecs[:, sort_idxs]
+  return evals, evecs
+
+
+def svd(A, precision=lax.Precision.HIGHEST):
```

Comment:
It looks to me like the polar decomposition is jittable, correct?

One thing we could probably do is hook up the polar decomposition to the API (because why not, it's useful) and use the existing `eigh` implementation as the basis of an SVD on TPU, which is simply missing at the moment.

We can separately work on making the `eigh` above jittable.

Reply:
Yes, it should be jittable